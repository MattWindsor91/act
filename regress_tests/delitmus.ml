(* This file is part of 'act'.

   Copyright (c) 2018, 2019 by Matt Windsor

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to permit
   persons to whom the Software is furnished to do so, subject to the
   following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
   NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
   USE OR OTHER DEALINGS IN THE SOFTWARE. *)

(** Tests the output of the 'act c delitmus' command. *)

open Core
module Ac = Act_common

let pp_cvars : Ac.C_variables.Map.t Fmt.t =
  Fmt.(
    prefix
      (unit "@,@,// C variables:@,")
      (vbox
         (using Ac.C_id.Map.keys
            (list ~sep:sp (hbox (prefix (unit "// -@ ") Ac.C_id.pp))))))

let pp_post : Act_c.Mini_litmus.Ast.Postcondition.t Fmt.t =
  Fmt.(
    prefix
      (unit "@,@,// Postcondition:@,")
      (hbox (prefix (unit "// ") Act_c.Mini_litmus.Pp.pp_post)))

let summarise_c_output (o : Act_c.Filters.Output.t) : unit =
  Fmt.(
    pr "@[<v>%a%a@]@." pp_cvars
      (Act_c.Filters.Output.cvars o)
      (option pp_post)
      (Act_c.Filters.Output.post o))

let delitmus_file ~(file : Fpath.t) ~(path : Fpath.t) : unit Or_error.t =
  ignore file ;
  Or_error.Let_syntax.(
    let%map output =
      Act_c.Filters.Litmus.run_from_fpaths Act_c.Filters.Delitmus
        ~infile:(Some path) ~outfile:None
    in
    summarise_c_output output)

let run (test_dir : Fpath.t) : unit Or_error.t =
  let full_dir = Fpath.(test_dir / "litmus" / "") in
  Common.regress_on_files "Delitmus" ~dir:full_dir ~ext:"litmus"
    ~f:delitmus_file

let command =
  Common.make_regress_command ~summary:"runs delitmusifier regressions" run

let () = Command.run command
