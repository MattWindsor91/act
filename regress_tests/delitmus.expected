# Delitmus tests

## iriw.c.litmus

```
// <!> Auto-generated from a litmus test by act.
#include <stdatomic.h>
atomic_int x = 0;
atomic_int y = 0;
int t2r0 = 0;
int t2r1 = 0;
int t3r0 = 0;
int t3r1 = 0;
void
P0()
{ atomic_store_explicit(&x, 1, memory_order_seq_cst); }
void
P1()
{ atomic_store_explicit(&y, 1, memory_order_seq_cst); }
void
P2()
{
    t2r0 = atomic_load_explicit(&x, memory_order_acquire);
    t2r1 = atomic_load_explicit(&y, memory_order_seq_cst);
}
void
P3()
{
    t3r0 = atomic_load_explicit(&y, memory_order_acquire);
    t3r1 = atomic_load_explicit(&x, memory_order_seq_cst);
}


// C variables:
// - t3r0
// - t3r1
// - x
// - y

// Postcondition:
// exists (2:r0 == 1 /\ 2:r1 == 0 /\ 3:r0 == 1 /\ 3:r1 == 0)
```
## iriw_fuzz_redundant.c.litmus

```
// <!> Auto-generated from a litmus test by act.
#include <stdatomic.h>
int MaJgQmON3yu = 2728;
int VFTQNWN = -13;
int ZgoV7uel9o = 1073;
int m02M5s = -64856;
atomic_int ne0J8wrPQam = -12197004;
atomic_int x = 0;
atomic_int y = 0;
int t3r0 = 0;
int t3r1 = 0;
int t4r0 = 0;
int t4r1 = 0;
void
P0()
{ atomic_store_explicit(&ne0J8wrPQam, 3417, memory_order_seq_cst); }
void
P1()
{
    atomic_store_explicit(&ne0J8wrPQam, 481486, memory_order_relaxed);
    atomic_store_explicit(&x, 1, memory_order_seq_cst);
}
void
P2()
{ atomic_store_explicit(&y, 1, memory_order_seq_cst); }
void
P3()
{
    t3r0 = atomic_load_explicit(&x, memory_order_acquire);
    t3r1 = atomic_load_explicit(&y, memory_order_seq_cst);
    atomic_store_explicit(&ne0J8wrPQam, -7, memory_order_seq_cst);
}
void
P4()
{
    t4r0 = atomic_load_explicit(&y, memory_order_acquire);
    t4r1 = atomic_load_explicit(&x, memory_order_seq_cst);
}


// C variables:
// - MaJgQmON3yu
// - VFTQNWN
// - ZgoV7uel9o
// - m02M5s
// - ne0J8wrPQam
// - t4r0
// - t4r1
// - x
// - y

// Postcondition:
// exists (2:r0 == 1 /\ 2:r1 == 0 /\ 3:r0 == 1 /\ 3:r1 == 0)
```
## memalloy_cmpxchg.c.litmus

```
// <!> Auto-generated from a litmus test by act.
#include <stdatomic.h>
atomic_int x = 0;
int t0r0 = 0;
void
P0()
{
    t0r0 = 0;
    atomic_compare_exchange_strong_explicit(&x, &t0r0, 2,
                                            memory_order_relaxed,
                                            memory_order_relaxed);
}
void
P1()
{ atomic_store_explicit(&x, 1, memory_order_relaxed); }


// C variables:
// - t0r0
// - x

// Postcondition:
// exists (0:r0 == 0 /\ x == 2)
```
## memalloy_na_int.c.litmus

```
// <!> Auto-generated from a litmus test by act.
#include <stdatomic.h>
int x = 0;
int t0r0 = 0;
void
P0()
{ x = 1; t0r0 = x; }


// C variables:
// - t0r0
// - x

// Postcondition:
// exists (0:r0 == 0 /\ x == 1)
```
## rwc.c.litmus

```
// <!> Auto-generated from a litmus test by act.
#include <stdatomic.h>
atomic_int x = 0;
atomic_int y = 0;
int t1r0 = 0;
int t1r1 = 0;
int t2r0 = 0;
void
P0()
{ atomic_store_explicit(&x, 1, memory_order_seq_cst); }
void
P1()
{
    t1r0 = atomic_load_explicit(&x, memory_order_acquire);
    t1r1 = atomic_load_explicit(&y, memory_order_seq_cst);
}
void
P2()
{
    atomic_store_explicit(&y, 1, memory_order_seq_cst);
    t2r0 = atomic_load_explicit(&x, memory_order_seq_cst);
}


// C variables:
// - t1r1
// - t2r0
// - x
// - y

// Postcondition:
// exists (1:r0 == 1 /\ 1:r1 == 0 /\ 2:r0 == 0)
```

Ran 5 test(s).
